backend/remote-state/pg/client.go

```diff
@@ -0,0 +1,181 @@
+package pg
+
+import (
+	"context"
+	"crypto/md5"
+	"database/sql"
+	"fmt"
+
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	_ "github.com/lib/pq"
+)
+
+// RemoteClient is a remote client that stores data in a Postgres database
+type RemoteClient struct {
+	Client     *sql.DB
+	Name       string
+	SchemaName string
+
+	// In-flight database transaction. Empty unless Locked.
+	txn  *sql.Tx
```

Comment:
we're going to want a mutex in here to protect these in the case of concurrent calls.

Reply:
Good catch. Mutex is now in-place to ensure sanity.